lambda/src/new_account_iam_role.py

```diff
@@ -0,0 +1,383 @@
+    return role
+
+
+def iam_role_create_policy(iam_client, role, role_name, policy_arn):
+    """Return True if permission policy can be attached, else False."""
+    LOG.info("%s: Attaching policy %s", role_name, policy_arn)
+    is_success = True
+    try:
+        role.attach_policy(PolicyArn=policy_arn)
+        iam_add_role_description(iam_client, role_name)
+        role.reload()
+    except (
+        botocore.exceptions.ClientError,
+        botocore.exceptions.ParamValidationError,
+        KeyError,
+    ) as exc:
+        # Note:  KeyError is seen when the policy ARN has a invalid policy
+        # name, e.g. isn't a known policy name.
+        LOG.error("%s: Unable to attach policy:\n\t%s", role_name, exc)
+        is_success = False
+    return is_success
+
+
+def main(
+    aws_profile,
+    role_name,
+    role_permission_policy,
+    trust_policy_json,
+    assume_role_arn=None,
+    partition="aws",
+    botocore_cache_dir=BOTOCORE_CACHE_DIR,
+):  # pylint: disable=too-many-arguments
+    """Create or update a IAM role with a trusted relationship."""
+    # Validate trust policy contains properly formatted JSON.  This is
+    # not a validation against a schema, so the JSON could still be bad.
+    try:
+        json.loads(trust_policy_json)
+    except json.decoder.JSONDecodeError as exc:
+        # pylint: disable=raise-missing-from
+        raise IamRoleInvalidArgumentsError(
+            f"'trust-policy-json' contains badly formed JSON:"
+            f"\n\t{exc}\n\tJSON input:  {trust_policy_json}"
+        )
+
+    # Validate that either role arn or an AWS profile was supplied, as one
+    # of them is needed to create a AWS session.
+    if not assume_role_arn and not aws_profile:
+        raise IamRoleInvalidArgumentsError(
+            "One of 'assume-role-arn' or 'aws-profile' is required"
+        )
+
+    # Create a session using the role arn or AWS profile.
+    session = get_session(assume_role_arn, aws_profile, botocore_cache_dir)
+    iam_resource = session.resource("iam")
+    iam_client = session.client("iam")
+
+    # Create a role using the role name and assign it an assumed trust policy
+    # with the user-supplied JSON.
+    role = iam_role_create_trust(iam_resource, iam_client, role_name, trust_policy_json)
+    if not role:
+        raise IamRoleInvalidArgumentsError(f"Unable to create '{role_name}' role.")
+
+    # Attach the permission policy(s) associated with the role.
+    policy_arn = f"arn:{partition}:iam::aws:policy/{role_permission_policy}"
+    if not iam_role_create_policy(iam_client, role, role_name, policy_arn):
+        raise IamRoleInvalidArgumentsError(
+            f"Unable to attach '{policy_arn}'to {role_name}."
+        )
+    return 0
+
+
+def lambda_handler(event, context):  # pylint: disable=unused-argument
+    """Entry point if script called by AWS Lamdba."""
+    LOG.info("Received event:\n%s", json.dumps(event))
```

Comment:
Didn't hear back from you re: my comment on the extra fields the injected decorator would add to the log, so I'm assuming you're ok with me not adding the decorator.  Keeping the log as is.

Reply:
If not using the decorator, then please use the `extras` feature.

Reply:
On second thought, I think using the decorator will be necessary. The information injected, in particular `function_request_id`, is necessary to query/filter the output for a given invocation of the lambda.

Reply:
I believe using the decorator impacts the tests -- there's more mocking involved.

I'd be nice to have requirements in advance because this feels like an endless game of whack-a-mole.